src/ErrorPage.php

```diff
@@ -0,0 +1,339 @@
+<?php
+/**
+ * Class ExtraThemeAndPluginHeaders.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AmpProject\AmpWP\Infrastructure\Service;
+use Exception;
+
+/**
+ * Produces an error page similar to wp_die().
+ *
+ * The actual wp_die() function cannot be used within the AMP response
+ * preparation code, as its 'exit' argument is only usable from WP 5.1 onwards.
+ *
+ * @package AmpProject\AmpWP
+ * @since   2.0.1
+ * @internal
+ */
+final class ErrorPage implements Service {
+
+	/**
+	 * Title of the error page.
+	 *
+	 * @var string
+	 */
+	private $title = '';
+
+	/**
+	 * Message of the error page.
+	 *
+	 * @var string
+	 */
+	private $message = '';
+
+	/**
+	 * Exception of the error page.
+	 *
+	 * @var Exception
+	 */
+	private $exception;
+
+	/**
+	 * Response code of the error page.
+	 *
+	 * @var int
+	 */
+	private $response_code = 500;
+
+	/**
+	 * Set the title of the error page.
+	 *
+	 * @param string $title Title to use.
+	 * @return self
+	 */
+	public function with_title( $title ) {
+		$this->title = $title;
+		return $this;
+	}
+
+	/**
+	 * Set the message of the error page.
+	 *
+	 * @param string $message Message to use.
+	 * @return self
+	 */
+	public function with_message( $message ) {
+		$this->message = $message;
+		return $this;
+	}
+
+	/**
+	 * Set the exception of the error page.
+	 *
+	 * @param Exception $exception Exception to use.
+	 * @return self
+	 */
+	public function with_exception( Exception $exception ) {
+		$this->exception = $exception;
+		return $this;
+	}
+
+	/**
+	 * Set the response_code of the error page.
+	 *
+	 * @param int $response_code Response code to use.
+	 * @return self
+	 */
+	public function with_response_code( $response_code ) {
+		$this->response_code = $response_code;
+		return $this;
+	}
+
+	/**
+	 * Render the error page.
+	 *
+	 * This first sets the required headers and then returns the HTML to send as
+	 * output.
+	 *
+	 * @return string
+	 */
+	public function render() {
+		$this->send_to_error_log();
+
+		$this->set_headers();
+
+		$text_direction = function_exists( 'is_rtl' ) && is_rtl()
+			? 'rtl'
+			: 'ltr';
+
+		$styles = $this->get_styles( $text_direction );
+		$html   = $this->get_html( $styles, $text_direction );
+
+		return $html;
+	}
+
+	/**
+	 * Send the exception that was caught to the error log.
+	 */
+	private function send_to_error_log() {
```

Comment:
I think this should short-circuit if the exceptions aren't normally sent to the error log:

```suggestion
	private function send_to_error_log() {
		if ( ! ( error_reporting() & E_ERROR ) ) { // And E_USER_ERROR & E_RECOVERABLE_ERROR?
			return;
		}
```

Some hosts may turn off error logging by default to to avoid filling up logs. So we should respect that.

Reply:
I think `E_USER_ERROR` and `E_RECOVERABLE_ERROR` can be ignored in this case.